Working draft of NHACP 0.1, for review and iteration.

thorpej opened the pull request.

thorpej committed:
Update nabu-comms.md

* Define protocol versioning.
* Define initial set of error codes.

thorpej committed:
Update nabu-comms.md

Remove EEOF -- I got a little out over my skis on that one.

thorpej committed:
Define a new START-NHACP message that does not collide with any other
known NABU messages and has protocol version information and a distinct
signature making future collisions less likely.

thorpej committed:
Change MTU to 8256 bytes, and provide rationale for this choice.

thorpej commented:
N.B. this is not yet ready to merge.  Changes still trickling in.

GryBsh commented on nabu-comms.md:
```diff
@@ -49,14 +49,27 @@ messages until the NABU requests to see them.
+transmitted within one second, there is a hard ceiling on the maximum
+transmission unit, dictacted by the NABU hardware.  With the native
+tramission rate of 111860 bits per second and many network adapter
+implementations using 2 stop bits, 11186 bytes is the practical limit
```
I believe this number is even lower, 10,169. 1 start bit, 8 data bits, and 2 stop bits is 11 bits. So the maximum throughput in real data is 8 out of 11 bits or 72.72% of 111860bps which is the same as 111860bps/11b=10169Bps.

thorpej replied:
I always forget the START bit.  Thanks!

GryBsh replied:
Everybody needs a start bit, but nobody remembers them 😄

thorpej committed:
Document GET-ERROR-DETAILS.

thorpej committed:
Review feedback: The practical limit is actually 10169 bytes/sec
(1 start, 8 data, 2 stop -> 11 bits per byte).

thorpej committed:
Cap STORAGE-GET and STORAGE-PUT at 8192 bytes.

thorpej committed:
Define STORAGE-GET-BLOCK and STORAGE-PUT-BLOCK.

thorpej committed:
Move the protocol version definitions into the Protocol Versioning
section.

thorpej committed:
Minor wording tweaks.

thorpej committed:
Add some example exchanges.

thorpej committed:
Define open flags.

thorpej committed:
Add an anticipatory ENOTDIR error code.

thorpej committed:
Add a 16-bit options field to START-NHACP.  No option bits yet defined.

thorpej committed:
Add DIR-LIST and DIR-GET-ENTRY requests and DIR-ENTRY response to support
directory enumeration.

thorpej committed:
Re-factor the DATE-TIME response and file attributes in DIR-ENTRY response
into DATE-TIME and FILE-ATTRS aggregate types that can be includes in
other payloads.

thorpej committed:
DIR-LIST -> LIST-DIR, DIR-GET-ENTRY -> GET-DIR-ENTRY

thorpej committed:
Rename STORAGE-CLOSE to FILE-CLOSE.  The semantics and behavior are
unchanged.

thorpej committed:
Use the term "file descriptor" instead of "storage slot".

thorpej committed:
Add FILE-READ, FILE-WRITE, FILE-SEEK requests.  Add UINT8-VALUE, UINT16-VALUE,
and UINT32-VALUE responses.  Add some clarifying language around seekability.

thorpej committed:
Tweak wording around FILE-READ and FILE-WRITE to reference file descriptors,
rather than storage.

thorpej committed:
Don't define the ERROR response with a fixed message size.

thorpej committed:
Add an empty options field in the START-NHACP example exchange.

thorpej committed:
Add optional CRC-8/WCDMA frame error check.

thorpej committed:
Fix wording botch.

thorpej committed:
Fix another instance of bad wording.

thorpej committed:
Per discussion w/ tergav17, allow STORAGE-GET-BLOCK to perform
partial reads (resulting buffer is zero-padded) and extending
writes (semantics are identical to STORAGE-PUT).

thorpej committed:
Fix typo.

thorpej committed:
Codify Postel's Law with regards to message processing.

thorpej committed:
Add some verbiage about seekability and writability and how implementations
are allowed to handle those.  Add some priority ordering to related error
returns.

thorpej committed:
Add REMOVE and RENAME operations.  Fix omission in STORAGE-PUT-BLOCK.

thorpej committed:
Document the new request framing / mutiplexing scheme, and tighten
up various wording.

thorpej committed:
Update for new O_RDWP flag, and re-arrange flags to group the access
mode together.  Tighten language around when the adapter needs to
shadow-buffer and what the behavior should be.

thorpej committed:
Improve the wording around O_DIRECTORY.

thorpej committed:
Tweak the wording around O_RDONLY, O_RDWR, and O_RDWP, and eliminate
the O_ACCMODE definition (merely describe that the least significant
3 bits form an access mode enumeration).

thorpej committed:
Add O_TRUNC flag.

thorpej committed:
Add FILE-GETATTR, FILE-SETSIZE, and MKDIR requests, and FILE-ATTRS
response.  Shuffle a couple of request types to group them better.

thorpej committed:
Note that this protocol is not strictly dependent on serial
communication interfaces.  Any stream-oriented transport will do.

thorpej committed:
- Rename FILE-GETATTR to FILE-GET-INFO.
- Rename the DIR-ENTRY response to FILE-INFO and share it for both
  GET-DIR-ENTRY and FILE-GET-INFO.  For the latter, the returned
  name-length is always 0.

thorpej committed:
FILE-SETSIZE -> FILE-SET-SIZE.

thorpej committed:
Fix typo.

thorpej committed:
Define a STRING aggregate type that's backwards-compatible with the
previous open-coded strings in various messages (Pascal-like), but
also optionally have early C-like 0-termination when sent by clients.

Redefine the rest of the messages that used strings to use STRING.

thorpej committed:
Note addition of STRING.

thorpej committed:
Tweak LIST-DIR; a 0-length pattern means "list entire directory".

thorpej committed:
In STORAGE-OPEN, specify that a zero-length URL string and the
O_DIRECTORY flag will open the root directory.

This allows a sequence of STORAGE-OPEN("") LIST-DIR("") to list the
root directory of the storage area.

thorpej committed:
Some formatting and consistency tweaks:
- Use "file descriptor" everywhere (no more "storage index" or
  "slot").  Also adjust the name of the fields to reflect this.
- Fix some table formatting.
- Insert horizontal rules around the request message section and
  the response message sections to better provide some visual
  separation.

thorpej committed:
Specify CRC-8/CDMA2000 rather than CRC-8/WCDMA, per request from
John Winans; it will be easier to implement in hardware for this
application.
